File: haskell/app/Game/Draw.hs
module Game.Draw where

import Control.Monad
import Data.Array
import Data.Array.Storable

import Canvas
import Game.Field

type Point = (Int, Int)

field :: Field -> Point -> Int -> Bool -> Canvas -> IO ()
field f (y0, x0) cellPx showDisappearing canvas = do
  drawBorder canvas
  drawCells canvas
  where
    drawBorder :: Canvas -> IO ()
    drawBorder canvas =
      forM_ [((y, x), white) | y <- [y0..y1], x <- [x0..x1], isBorder y x] $
        uncurry $ writeArray canvas
      where
        isBorder y x = y `elem` [y0, y1] || x `elem` [x0, x1]

        (y1, x1) = (y0 + h * cellPx, x0 + w * cellPx)
          where
            (h, w) = dims f

    drawCells :: Canvas -> IO ()
    drawCells canvas =
      forM_ drawableCells $
        \(Pos (y, x), c) -> cell (cellColor c) (y0 + y * cellPx, x0 + x * cellPx) cellPx canvas
      where
        drawableCells = filter isDrawable $ assocs f
          where
            isDrawable (_, c) = case c of
              Falling      -> True
              Frozen       -> True
              Disappearing -> showDisappearing
              Empty        -> False

        cellColor Falling = blue
        cellColor _       = white

cell :: Pixel -> Point -> Int -> Canvas -> IO ()
cell color (y, x) cellPx canvas = do
  forM_ (zip [cellPx, cellPx - 2, cellPx - 8] $ cycle [black, color]) $
    \(px, color) -> fillSquare color (y + cellPx - px, x + cellPx - px) (y + px - 1, x + px - 1) canvas

fillSquare :: Pixel -> Point -> Point -> Canvas -> IO ()
fillSquare color (y0, x0) (y1, x1) canvas =
  forM_ [((y, x), color) | y <- [y0..y1], x <- [x0..x1]] $
    uncurry $ writeArray canvas

clear :: Canvas -> IO ()
clear canvas = do
  (b, e) <- getBounds canvas
  fillSquare black b e canvas

{-
TODO: why this won't work (recursive) but forM_ does?

cell :: Pixel -> Point -> Int -> Canvas -> IO ()
cell color (y, x) cellPx canvas = do
    go color [cellPx] --[cellPx, cellPx - 2, cellPx - 8]
    where 
        go :: Pixel -> [Int] -> IO ()
        go color []         = pure ()
        go color (px : pxs) = do
            print (length pxs)
            let color = current color
            print ((x + cellPx - px, y + cellPx - px), (x + px, y + px))
            putStrLn "begin fill"
            fillSquare color (x + cellPx - px, y + cellPx - px) (x + px, y + px) canvas
            putStrLn "end fill"
            go color pxs
                where
                    current color = if color == background then white else black
                    background    = black

fillSquare :: Pixel -> Point -> Point -> Canvas -> IO ()
fillSquare color (x0, y0) (x1, y1) canvas = do
    putStrLn "begin go"
    go [((y, x), color) | y <- [y0..(y1 - 1)]
                        , x <- [x0..(x1 - 1)] ]
    putStrLn "end go"
        where
            go :: [((Int, Int), Pixel)] -> IO ()
            go []                = pure ()
            go ((i, color) : xs) = do
                print i
                putStrLn "begin writeArray"
                writeArray canvas i color
                putStrLn "end writeArray"
                go xs
-}

File: rust/src/game/draw.rs
use std::iter::zip;

use itertools::{chain, iproduct};

use crate::game::RawCanvas;

use super::field::Cell;
use super::grid::Grid;

pub fn grid<const N: usize, const M: usize>(
    canvas: &mut Canvas,
    bucket: &Grid<Cell, N, M>,
    cell_px: usize,
    [y0, x0]: [usize; 2],
    show_disappearing: bool,
) {
    // draw border
    {
        let y1 = y0 + bucket.height() * cell_px - 1;
        let x1 = x0 + bucket.width()  * cell_px - 1;

        for (y, x) in chain(iproduct!(y0..=y1, [x0, x1]), iproduct!([y0, y1], (x0..=x1))) {
            if let Some(pixel) = canvas.get_mut(x, y) {
                *pixel = color::WHITE;
            }
        }
    }

    let is_drawable = |c| c == Cell::Falling || c == Cell::Frozen || (c == Cell::Disappearing && show_disappearing);
    let to_color = |c| if c == Cell::Falling { color::BLUE } else { color::WHITE };
    bucket.iter_with_idx()
        .filter(|(_, &c)| is_drawable(c))
        .map(|(i, &c)| (i, to_color(c)))
        .for_each(|([y, x], color)| {
            self::cell(canvas, cell_px, [y0 + y * cell_px, x0 + x * cell_px], color);
        });
}

pub fn cell(canvas: &mut Canvas, cell_px: usize, [y0, x0]: [usize; 2], color: u32) {
    let borders = [cell_px, cell_px - 2, cell_px - 8];
    let colors = [color::BLACK, color].into_iter().cycle();

    for (px, color) in zip(borders, colors) {
        fill_square(canvas, color, [y0 + cell_px - px, x0 + cell_px - px], [y0 + px, x0 + px]);
    }
}

pub mod color {
    pub const BLACK: u32 = 0x0000_0000;
    pub const WHITE: u32 = 0x00FF_FFFF;
    pub const GREEN: u32 = 0x0000_FF00;
    pub const BLUE: u32 = 0x0000_00FF;
}

fn fill_square(canvas: &mut Canvas, color: u32, [y0, x0]: [usize; 2], [y1, x1]: [usize; 2]) {
    for y in y0..y1 {
        for x in x0..x1 {
            if let Some(pixel) = canvas.get_mut(x, y) {
                *pixel = color;
            }
        }
    }
}

pub fn clear(canvas: &mut Canvas) {
    fill_square(canvas, color::BLACK, [0, 0], [canvas.height(), canvas.width()]);
}

pub struct Canvas<'a> {
    inner: &'a mut dyn RawCanvas,
}

impl<'a> Canvas<'a> {
    pub fn from_raw(inner: &'a mut dyn RawCanvas) -> Self {
        Self {
            inner,
        }
    }
}

impl Canvas<'_> {
    pub fn width(&self) -> usize {
        self.inner.width()
    }

    pub fn height(&self) -> usize {
        self.inner.height()
    }

    pub fn get_mut(&mut self, x: usize, y: usize) -> Option<&mut u32> {
        let width = self.inner.width();
        let height = self.inner.height();

        if x < width && y < height {
            Some(&mut self.inner[y*width + x])
        } else {
            None
        }
    }
}